### PR TITLE
Permissions to perform actions in Conferences

### PR DESCRIPTION
**Pull request by LauraJaimeC**
- Allow to perform all actions in Conferences.
- Add columns: Department/Area, Department role, Conference role

### CODE REVIEW

File: db/migrate/20210420143021_add_area_to_conferences.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddAreaToConferences < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :decidim_conferences, :decidim_area, index: true
```
Comment (tramuntanal) [inline db/migrate/20210420143021_add_area_to_conferences.rb:5]: only `if defined?(Decidim::Conferences)`

File: decidim-department_admin.gemspec [hunk 2]
```diff
@@ -21,6 +21,7 @@ Gem::Specification.new do |s|
   s.files = Dir["{app,db,config,lib}/**/*", "LICENSE-AGPLv3.txt", "Rakefile", "README.md"]
   s.test_files = Dir["spec/**/*"]
 
+  s.add_dependency "decidim-conferences", DECIDIM_VER
```
Comment (tramuntanal) [inline decidim-department_admin.gemspec:24]: This dependency is optional, it probably should be an `add_development_dependency` instead. We should not obligate all users of `department_admin` to install `decidim-conferences`.

File: lib/decidim/department_admin/engine.rb [hunk 3]
```diff
@@ -65,6 +65,19 @@ class Engine < ::Rails::Engine
...
+        # **
+        require "decidim/conferences/admin/conference_copies_controller"
+        require "decidim/conferences/admin/components_controller"
+        artifact = ::Decidim::Conferences::Admin::Concerns::ConferenceAdmin
+        ConferencesAdminConcernPermissions = Class.new(::Decidim::DepartmentAdmin::Permissions)
+        register_new_permissions_for(artifact, ConferencesAdminConcernPermissions)
+
+        artifact = Decidim::Conferences::Admin::ApplicationController
+        ConferencesAdminApplicationControllerPermissions = Class.new(::Decidim::DepartmentAdmin::Permissions)
+        register_new_permissions_for(artifact, ConferencesAdminApplicationControllerPermissions)
```
Comment (tramuntanal) [inline lib/decidim/department_admin/engine.rb:80]: all this code between an `if defined?(Decidim::Conferences)` block

File: lib/decidim/department_admin/engine.rb [hunk 4]
```diff
@@ -82,6 +95,9 @@ class Engine < ::Rails::Engine
         # override assemblies space manifest permissions with DepartmentAdmin's one
         manifest = Decidim.find_participatory_space_manifest(:assemblies)
         manifest.permissions_class_name = "Decidim::Assemblies::ParticipatorySpacePermissions"
+        # override conferences space manifest permissions with DepartmentAdmin's one
+        manifest = Decidim.find_participatory_space_manifest(:conferences)
+        manifest.permissions_class_name = "Decidim::Conferences::ParticipatorySpacePermissions"
```
Comment (tramuntanal) [inline lib/decidim/department_admin/engine.rb:100]: all this code between an `if defined?(Decidim::Conferences)` block

**Review by tramuntanal — CHANGES_REQUESTED** [review]
Good job  @LauraJaimeC !
We should to take into account that `decidim-conferences` is optional, I've commented at the places were I think we should add a conditional.

File: app/decorators/decidim/admin/users_controller_decorator.rb [hunk 5]
```diff
@@ -68,6 +68,24 @@ def show
...
+      @spaces.push("title" => conference_title,
+                   "type" => t("models.user.fields.conference_type", scope: "decidim.admin"),
+                   "area" => area_name,
+                   "created_at" => conference.created_at,
+                   "private" => conference.private_space?,
+                   "published" => conference.published?)
+    end
+
     # rubocop: disable Style/NestedTernaryOperator
     if params[:sort_column] && (params[:sort_column] == "published" || params[:sort_column] == "private") && params[:sort_order] && params[:sort_order] == "asc"
```
[the remark this answers is not in the document]
Comment (LauraJaimeC) [reply app/decorators/decidim/admin/users_controller_decorator.rb:90]: I have not made changes in that code, even so, do you think it's ok to change it in this PR?

File: app/helpers/decidim/department_admin/application_helper.rb [hunk 6]
```diff
@@ -37,6 +41,14 @@ def user_assemblies_filtered(user, _locale, search_text)
         query = query.where("lower(title->>?) like lower(?)", current_locale, "%#{search_text}%") if search_text.present?
         query
       end
+
+      def user_conferences_filtered(user, _locale, search_text)
+        return unless defined?(Decidim::Conferences)
```
Comment (tramuntanal) [inline app/helpers/decidim/department_admin/application_helper.rb:46]: sorry I think returning `nil` will produce a null exception in line 24. Better return an empty array:
```suggestion
        return [] unless defined?(Decidim::Conferences)
```

File: app/queries/decidim/admin/user_admin_filter.rb [hunk 7]
```diff
@@ -48,28 +48,58 @@ def filter_by_search_text(users)
...
-                    where lower(title->>?) like lower(?)))
-          or id in  ( select decidim_user_id
-                      from decidim_assembly_user_roles
-                      where decidim_assembly_id in (
-                        select id
-                        from decidim_assemblies
-                        where lower(title->>?) like lower(?))))
-        EOSQL
+
+        if defined?(Decidim::Conferences)
```
Comment (tramuntanal) [inline app/queries/decidim/admin/user_admin_filter.rb:52]: better avoid having this if-else structure, what will we do when we have to consider another space, like initiatives?
Can we construct the sql string before invoking the `where` and conditionaly concatenate the conferences part?

**Review by tramuntanal — CHANGES_REQUESTED** [review]

**Review by tramuntanal — DISMISSED** [review]
:+1:

File: app/decorators/decidim/admin/users_controller_decorator.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (tramuntanal) [reply app/decorators/decidim/admin/users_controller_decorator.rb:88]: Refactoring to simplify code should be always welcomed

**Review by tramuntanal — APPROVED** [review]

**Review by tramuntanal — APPROVED** [review]

**Pull request merged**